File: IN1000/Obliger/Oblig8/rutenett.py
import random
from celle import Celle

"""

Danner et "rutenett" objekt, som egentlig er en liste med lister.
Fyller hver posisjon i hver liste med Celler og har en del metoder for
å sjekke "naboene" til rutene. 

"""




# 1. Konstrukør ----------------------------------------------------------------


class Rutenett:


    def __init__(self, rader, kolonner):
        self._ant_rader = rader
        self._ant_kolonner = kolonner
        self._rutenett = self._lag_tomt_rutenett()
        #              ^ rutenett = kall på funksjon _lag_tomt_rutenett.




# 2. Rutenett konstruksjon -----------------------------------------------------


    #Lager "rutenett" ved å danne en liste med lister.
    def _lag_tomt_rutenett(self):

        rutenett = []


        for rad in range(self._ant_rader): # kjører _ant_rader ganger
            rutenett.append(self._lag_tom_rad())
                            # ^legger til i rutenett et kall på funksjonen _lag_tom_rad


        return rutenett



    #lager ver enkelt kolonne med None som innhold i hver indeks.
    def _lag_tom_rad(self):

        kolonner = []

        #appender None til en liste _ant_kolonner ganger
        for kol in range(self._ant_kolonner):
            kolonner.append(None)

        return kolonner #returnerer listen til _lag_tomt_rutenett




# 3. Fylle celler i rutenett ---------------------------------------------------

    # Fyller alle plasser med en levende eller død celle.
    def fyll_med_tilfeldige_celler(self):



        # for løkke for hver rad med nøstet for løkke for hver for kolonne.
        # Gjør alle kordinater om til celler, der 1/3 av blir levende.


        # visuelt eksempel av koden:
        # A1 A2 A3 A4        <--     Kjører lag_celle på A1, B1, C1, D1
        # B1 B2 B3 B4                Så A2, B2, C2, D2 ...
        # C1 C2 C3 C4
        # D1 D2 D3 D4





        for rad in range(self._ant_rader): # for rad in _ant_rader ganger

            for kolonne in range(self._ant_kolonner): # for kolonne i _ant_kolonner ganger
                self.lag_celle(rad, kolonne)
                # ^ kall på lag celle med nåværende verdi av rad og kolonne som "kordinater"



    # Generer celle på oppgitt kordinat. 1/3 av dem får status levende.
    def lag_celle(self, rad, kol):
        self._rutenett[rad][kol] = Celle() #posisjon = nytt celle objekt

        if random.randint(0,2) == 2: # generer tilfeldig tall mellom 0 og 2
            self._rutenett[rad][kol].sett_levende() # sett celle som levende om tall == 2.




# 4. Naboer og tegning av rutenett ---------------------------------------------


    # Returnerer celle objekt ved oppgitte kordinater.
    def hent_celle(self, rad, kol):

        if rad < 0 or rad >= self._ant_rader:
            return None

        elif kol < 0 or kol >= self._ant_kolonner:
            return None


        # ^ retunerer None om kordinatene er utenfor rutenettet-

        else:
            return self._rutenett[rad][kol]
            # ^ Returnerer riktig objekt ved oppgitte "kordinater".



    # printer til terminal tegn for hver "rute" i rutenettet.
    # gir en visuell represtenasjon av det som egentlig er nøstede lister.
    def tegn_rutenett(self):

        for i in range(2):
            print()
        # ^ gir luft til koden



        for rad in range(self._ant_rader):

            for kolonne in range(self._ant_kolonner):

        #^ går en rute om gangen

                print(self.hent_celle(rad, kolonne).hent_status_tegn(), end="")
                # ^ printer tegn for statusen til cellen for hver rute.      ^fjerner linjeskifter fra enden.

            print("") #nytt linjeskift etter hver rad er skrevet ut.

        for i in range(2):
            print()
        # ^ luft i terminalen.



    # Sjekker de 8 rutene rundt valgt rute, og legger dem til i cellens instansvariabel.
    def _sett_naboer(self, rad, kol):


        celle = self.hent_celle(rad, kol) # lagrer referanse til celle.


        for offset_rad in range(-1, 2): #kolonner -1, 0, 1
            for offset_kol in range(-1, 2): # rader -1, 0, 1

                nabo = self.hent_celle(rad + offset_rad, kol + offset_kol)
                # ^ henter nabo ved å bruke nåværende posisjon og legge til en "offset" på +1, 0 eller -1

                if offset_rad == 0 and offset_kol == 0:
                    pass
                # ^ passer på at den ikke legger til seg selv.

                elif nabo != None: #hent_celle gir None om oppgit celle er utenfor rutenettet.
                    celle.legg_til_nabo(nabo) #legger til nabo i cellens _naboer instansvariabel.


    # kjører _sett_naboer på alle rutene i rutenettet.
    def koble_celler(self):

        for rad in range(self._ant_rader):

            for kolonne in range(self._ant_kolonner):
                self._sett_naboer(rad,kolonne)

        # Rad og kolonne endrer seg i forhold til hvor den er i for loopen.
        # Der rad og kolonne er et heltall som kan brukes til å danne et kordinat.



    # Samler alle cellene i en lang liste. (istedenfor et rutenett)
    def hent_alle_celler(self):

        alle_instanser = []


        for rad in self._rutenett:

            for kolonne in rad:
                alle_instanser.append(kolonne)
                # appender hver celle til en helt ordinær liste.

        return alle_instanser



    # Teller hvor mange levende celler det totalt er i rutenettet
    def antall_levende(self):


        antall_levende = 0


        for celle in self.hent_alle_celler():

                if celle.er_levende():
                    antall_levende += 1

                # sjekker om hvert objekt er levende, og teller den om den er levende.


        return antall_levende

File: IN1010/Oblig/oblig1/Rutenett.java
class Rutenett {


    // oppretter instansvariablene for rutenettet aka nøstede arrays
    public int antRader;
    public int antKolonner; // public, brukes i verdensprogram. kunne eventuelt laget ny metode for å returnere verdiene.
    private Celle[][] rutene;
    

    // Setter verdiene til valgte verdier. 
    public Rutenett(int rader, int kolonner) {

        antRader = rader; 
        antKolonner = kolonner;
        rutene = new Celle[antRader][antKolonner];
    }


    // oppretter nytt celle objekt i minnet og lagrer en referanse til den i valgt index i rutenett
    // Setter levende 1/3 av gangene.
    public void lagCelle(int rad, int kol) {

        Celle cellen = new Celle();
        rutene[rad][kol] = cellen;

        if (Math.random()<=0.3333) {
            cellen.settLevende();
        }
    }

    // kjører lagCelle() på alle celler i rutenettet. Resultatet er et rutenett fylt av celler, der 1/3 er levende. 
    public void fyllMedTilfeldigeCeller(){

        for (int rad = 0; rad < antRader; rad++) {
            for (int kol = 0; kol < antKolonner; kol ++) {
                lagCelle(rad, kol);
            }
        }
    }

    // henter ut celleobjektet av rutenettet. Returnerer null om opgitt index er utenfor rutenettet. 
    public Celle hentCelle(int rad, int kolonne) {

        if (rad >= antRader || rad < 0 || kolonne >= antKolonner || kolonne < 0) {
            return null;
        }      
        return rutene[rad][kolonne];
    }

    // Kjører hentStatusTegn() på alle celler i rutenettet og skriver ut resultatet i terminalen. 
    public void tegnRutenett(){


         for (int rad = 0; rad < antRader; rad++) {
            for (int kol = 0; kol < antKolonner; kol ++) {
                
                System.out.print(" " + rutene[rad][kol].hentStatusTegn() + " ");

            }
        System.out.println(" ");
    }
    }



    // Sjekker de 8 rutene rundt en valgt celle og lagrer en referanse til dem i cellens interne array. 
    public void settNaboer(int rad, int kol) {

        Celle hoved = rutene[rad][kol];

        for (int offset_rad = -1; offset_rad < 2; offset_rad++) {
            for (int offset_kol = -1; offset_kol < 2; offset_kol++) {

                int ny_r = offset_rad + rad;
                int ny_k = offset_kol + kol;

                
                
                // if hentCelle ikke er 0 på valgt index i rutenett AND not seg selv:
                if (hentCelle(rad + offset_rad, kol + offset_kol) != null && !(offset_rad == 0 && offset_kol == 0)){

                    
                    Celle nabo = hentCelle(ny_r, ny_k);
                    hoved.leggTilNabo(nabo);
                    
                } 
                }


        }
    }


    // kjører settNaboer på alle celler i rutenettet
    public void kobleAlleCeller() {

        for (int rad = 0; rad < antRader; rad++) {
            for (int kol = 0; kol < antKolonner; kol++) {

                settNaboer(rad, kol);
            }
        }
    }


    // teller antall celler med levende status i rutenettet, returneres som int
    public int antallLevende() {

        int teller = 0;

        for (int rad = 0; rad < antRader; rad++) {
            for (int kol = 0; kol < antKolonner; kol++) {

                if (rutene[rad][kol].erLevende()) {
                    teller++;
                }
            }
        }

        return teller;
}



}
